File: dsa/typescript/linkedlist.ts
class List<T>{
	private head:ListNode<T>;

	constructor(head:ListNode<T>){
		this.head = head;
	}

	public getHead():ListNode<T>{
		return this.head;
	}

	public addValue(value:T){
		let node = new ListNode(value, null);

		let head = this.head;

		while(head.getNext() != null){
			head = head.getNext();
		}

		head.setNext(node);
		return;
	}

	public addNode(node:ListNode<T>){
		let head = this.head;

		while(head.getNext() != null){
			head = head.getNext();
		}

		head.setNext(node);
		return;
	}

	public printNodes(){
		let head = this.head;
		while(head != null){
			console.log(head.getValue());

			head = head.getNext();
		}
		return;
	}

	// Iterative
	public getLength():number{
		let head = this.head;
		let count:number = 0;

		while(head != null){
			count += 1;
			head = head.getNext();
		}
		return count;
	}

	// Recursive

	public getMiddleNode():ListNode<T>{
		let mid = Math.floor(this.getLength() / 2 );
		let head = this.head;
		let counter:number = 0;

		while(head != null && counter < mid){
			counter += 1;
			head = head.getNext();
		}
		return head;
	}

	public getTail():ListNode<T>{
		let head = this.head;

		while(head.getNext() != null){
			head = head.getNext();
		}
		return head;
	}

	public addNodeAt(node:ListNode<T>, pos:number){
		let head = this.head;
		let count = 1;

		while(head.getNext() != null && count < pos){
			head = head.getNext();
			count++;
		}

		node.setNext(head.getNext());
		head.setNext(node);
		return;
	}

	// Delete node
	public deleteNode(node:ListNode<T>):boolean{
		let curr = this.head;
		let prev = null;

		while(curr != null && curr.getValue() != node.getValue()){
			prev = curr;
			curr = curr.getNext();
		}

		if(curr.getValue() == node.getValue()){
			prev.setNext(curr.getNext());
			return true;
		}
		if(curr == null){
			return false;
		}
	}

	public deleteByPosition(pos:number):boolean{
		let curr = this.head;
		let prev = null;
		let counter = 1;

		while(curr != null && counter < pos){
			counter += 1;

			prev = curr;
			curr = curr.getNext();
		}

		if(curr == null){
			return false;
		}

		prev.setNext(curr.getNext());
		return true;
	}

	public findNode(node:ListNode<T>):boolean{
		let curr = this.head;

		while(curr != null && curr.getValue() != node.getValue()){
			curr = curr.getNext();
		}

		if(curr == null){
			return false;
		}
		if(curr.getValue() == node.getValue()){
			return true;
		}
	}

	public findValue(value:T):boolean{
		let curr = this.head;

		while(curr != null && curr.getValue() != value){
			curr = curr.getNext();
		}

		if(curr == null){
			return false;
		}
		if(curr.getValue() == value){
			return true;
		}
	}
}

class ListNode<T>{

	private value:T;
	private next: ListNode<T>;

	constructor(value: T, next:ListNode<T>){
		this.value = (value === undefined ? null : value);
		this.next = (next === undefined ? null :  next);
	}

	public setValue(value: T){
		this.value = value;
	}

	public getValue():T{
		return this.value;
	}

	public getNext():ListNode<T>{
		return this.next;
	}

	public setNext(node:ListNode<T>){
		this.next = node;
	}
}

let head = new ListNode(100, null);
let list = new List(head);

let node1 = new ListNode(200, null);
let node2 = new ListNode(300, null);
let node3 = new ListNode(400, null);
let node4 = new ListNode(500, null);
let node5 = new ListNode(600, null);
let node6 = new ListNode(700, null);
let node7 = new ListNode(800, null);
let testNode = new ListNode(1000, null);

list.addNode(node1);
list.addNode(node2);
list.addNode(node3);
list.addNode(node4);
list.addNode(node5);
list.addNode(node6);

let ans = list.findNode(testNode);
console.log(ans);

File: dsa/go/linked list/LinkedList.go
package main

import "fmt"

// NODE ***********

type Node struct {
	Value int
	Next  *Node
}

func (n Node) getValue() int {
	return n.Value
}

func (n Node) setValue(value int) {
	n.Value = value
}

func (n Node) getNext() *Node {
	return n.Next
}

func (n Node) setNext(node *Node) {
	n.Next = node
}

// *********** LIST ***********

type List struct {
	Head *Node
}

func (l List) getHead() *Node {
	return l.Head
}

func (l List) getLength() int {
	head := l.getHead()
	count := 1
	for head.getNext() != nil {
		head = head.getNext()
		count++
	}

	return count
}

func (l List) getTail() *Node {
	head := l.getHead()
	for head.getNext() != nil {
		head = head.getNext()
	}

	return head
}

func (l List) addNode(n *Node) {
	head := l.getHead()

	for head.getNext() != nil {
		head = head.getNext()
	}

	fmt.Println("adding node to last")
	head.Next = n
	return
}

// *********** MAIN ***********

func main() {
	head := Node{
		Value: 100,
		Next:  nil,
	}

	node1 := Node{Value: 90, Next: nil}
	node2 := Node{Value: 80, Next: nil}
	node3 := Node{Value: 70, Next: nil}
	node4 := Node{Value: 60, Next: nil}
	node5 := Node{Value: 50, Next: nil}
	node6 := Node{Value: 40, Next: nil}
	node7 := Node{Value: 30, Next: nil}
	node8 := Node{Value: 20, Next: nil}

	list := List{Head: &head}

	head.Next = &node1
	node1.Next = &node2
	node2.Next = &node3
	node3.Next = &node4
	node4.Next = &node5
	node5.Next = &node6
	node6.Next = &node7
	node7.Next = &node8

	nodeTest := Node{Value: 2000, Next: nil}

	list.addNode(&nodeTest)
	fmt.Println(list.getTail().getValue())
}
